modeling/algorithm/local-shared-algorithm.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package algorithm
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+
+	"github.com/googleinterns/k8s-topology-simulator/modeling/types"
+	"k8s.io/klog/v2"
+)
+
+// LocalSharedSliceAlgorithm is one variation of LocalSliceAlgorithm which
+// 'borrows' and 'rents' endpoints from other zones to make the local
+// EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution). This variation deals with failed corner cases while having a
+// slightly different distribution policy when approximating float to int
+type LocalSharedSliceAlgorithm struct{}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy, zones with 0.xx expected endpoints while have no endpoints in
+// zone will be treated with a higher priority.
```

Comment:
Maybe something like this?
```suggestion
// zone' policy, zones with no endpoints allocated will be treated with a higher priority.
```

Reply:
changed